src/aof.c

```diff
@@ -815,38 +815,43 @@ int openNewIncrAofForAppend(void) {
 #define AOF_REWRITE_LIMITE_THRESHOLD    3
 #define AOF_REWRITE_LIMITE_MAX_MINUTES  60 /* 1 hour */
 int aofRewriteLimited(void) {
-    int limit = 0;
-    static int limit_delay_minutes = 0;
     static time_t next_rewrite_time = 0;
 
     unsigned long incr_aof_num = listLength(server.aof_manifest->incr_aof_list);
-    if (incr_aof_num >= AOF_REWRITE_LIMITE_THRESHOLD) {
+    if (incr_aof_num < AOF_REWRITE_LIMITE_THRESHOLD) {
+        /* the incr aof number possible change to be normal, such as a successful BGREWRITEAOF.
+         * so we need to reset the next_rewrite_time before return 'no limit' */
+        next_rewrite_time  = 0;
+        return 0;
+    }
+    
+    /* if it is in the limiting state, then check if the next_rewrite_time is reached */
+    if (next_rewrite_time != 0) {
         if (server.unixtime < next_rewrite_time) {
-            limit = 1;
+            return 1;
         } else {
-            if (limit_delay_minutes == 0) {
-                limit = 1;
-                limit_delay_minutes = 1;
-            } else {
-                limit_delay_minutes *= 2;
-            }
-
-            if (limit_delay_minutes > AOF_REWRITE_LIMITE_MAX_MINUTES) {
-                limit_delay_minutes = AOF_REWRITE_LIMITE_MAX_MINUTES;
-            }
-
-            next_rewrite_time = server.unixtime + limit_delay_minutes * 60;
+            next_rewrite_time = 0;
+            return 0;
+        }
+    }
 
-            serverLog(LL_WARNING,
-                "Background AOF rewrite has repeatedly failed %ld times and triggered the limit, will retry in %d minutes",
-                incr_aof_num, limit_delay_minutes);
+    int limit_delay_minutes = 1;
+    int delay_count = incr_aof_num - AOF_REWRITE_LIMITE_THRESHOLD;
+    while (delay_count) {
+        limit_delay_minutes *= 2;
```

Comment:
this loop can be replaced with `limit_delay_minutes <<= delay_count` (after clamping `delay_count` to some maximum)

Reply:
IIUC you may have misunderstood the meaning of this delay, when redis trigger the limit (ie incr_aof_num > AOF_REWRITE_LIMITE_THRESHOLD), we will increase our delay time strictly according to the
sequence of 1, 2, 4, 8, instead of simply using `incr_aof_num * 2` , In this way we will get a jumping discontinuous delay (If you use `bgrewriteaof` before the end of the next delay and generate a lot of failed incr aof).

Reply:
you mean the limit should be increased only on automatic attempts, and not on manual ones?
you're right that this is a key difference between the two methods, but i'm not sure it's an important concern.

i still don't understand what was the problem with the original approach here (i.e. why do this computation instead of the static var).
even if for some reason the `growth` ratio shrieked, the re-location of the call to this function could have been enough to solve that problem.

Reply:
Yes, but now I  am more inclined to this PR solution now, we should really determine our delay time by the number of real failures, rather than the number of automatic AOFRW failures (we need to put the manual AOFRW also counted)

Reply:
logically, i think automatic triggered failures is more correct, but considering counting real failures simplifies the code, i think i'm willing to change it since i don't think it's an important concern.
on the other hand, the code that relies on the static var is not that complex either..

Reply:
> 

@chenyang8094 oh, `delay time strictly` is also a good way, but using `incr_aof_num * 2` is simple for us to understand,  there is no key difference between them

Reply:
Another problem, `incr_aof_num` does not necessarily represent the number of failures. If  redis starts from a data set with many incr AOFs(may be a AOF backup). At this time, it does not have any AOFRW failures, but it still has to wait 60 minutes to execute the first AOFRW.